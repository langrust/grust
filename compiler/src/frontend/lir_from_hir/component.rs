use itertools::Itertools;

use crate::{
    common::r#type::Type,
    hir::{identifier_creator::IdentifierCreator, node::Node},
    lir::{
        expression::Expression as LIRExpression,
        item::state_machine::{
            event::{Event, EventElement},
            input::{Input, InputElement},
            state::{init::Init, step::Step, State},
            StateMachine,
        },
    },
    symbol_table::SymbolTable,
};

use super::LIRFromHIR;

impl LIRFromHIR for Node {
    type LIR = StateMachine;

    fn lir_from_hir(self, symbol_table: &SymbolTable) -> Self::LIR {
        let Node {
            id,
            statements,
            memory,
            contract,
            ..
        } = self;
        let mut identifier_creator = IdentifierCreator::from(vec![]);

        // get node name
        let name = symbol_table.get_name(id).clone();

        // get node inputs
        let mut inputs = symbol_table
            .get_node_inputs(id)
            .into_iter()
            .map(|id| {
                (
                    symbol_table.get_name(*id).clone(),
                    symbol_table.get_type(*id).clone(),
                )
            })
            .collect::<Vec<_>>();

        // create the event structure
        let event = if let Some(event_enum_id) = symbol_table.get_node_event_enum(id) {
            // add the event input
            let event_id = symbol_table.get_node_event(id).unwrap();
            inputs.push((
                symbol_table.get_name(event_id).clone(),
                Type::Enumeration {
                    name: symbol_table.get_name(event_enum_id).clone(),
                    id: event_enum_id,
                },
            ));

            // get enum elements
            let mut elements = symbol_table
                .get_event_enum_elements(event_enum_id)
                .iter()
                .map(|element_id| EventElement::InputEvent {
                    identifier: symbol_table.get_name(*element_id).clone(),
                    r#type: symbol_table.get_type(*element_id).clone(),
                })
                .collect::<Vec<_>>();

            // get generics
            let generics = elements
                .iter_mut()
                .flat_map(|event_elem| match event_elem {
                    EventElement::InputEvent { r#type, .. } => {
                        r#type.get_generics(&mut identifier_creator)
                    }
                    EventElement::NoEvent => vec![],
                })
                .collect::<Vec<_>>();

            Some(Event {
                node_name: name.clone(),
                elements,
                generics,
            })
        } else {
            None
        };

        // get node output type
        let outputs = symbol_table.get_node_outputs(id);
        let mut output_type = {
            let mut types = outputs
                .iter()
                .map(|(_, output_id)| symbol_table.get_type(*output_id).clone())
                .collect::<Vec<_>>();
            if types.len() == 1 {
                types.pop().unwrap()
            } else {
                Type::Tuple(types)
            }
        };

        // get node output expression
        let outputs = symbol_table.get_node_outputs(id);
        let output_expression = {
            let mut identifiers = outputs
                .iter()
                .map(|(_, output_id)| LIRExpression::Identifier {
                    identifier: symbol_table.get_name(*output_id).clone(),
                })
                .collect::<Vec<_>>();
            if identifiers.len() == 1 {
                identifiers.pop().unwrap()
            } else {
                LIRExpression::Tuple {
                    elements: identifiers,
                }
            }
        };

        // collect imports from statements, inputs and output types, memory and contracts
        let mut imports = statements
            .iter()
            .flat_map(|equation| equation.get_imports(symbol_table))
            .unique()
            .collect::<Vec<_>>();
        let mut inputs_type_imports = inputs
            .iter()
            .flat_map(|(_, typing)| typing.get_imports(symbol_table))
            .unique()
            .collect::<Vec<_>>();
        let mut output_type_imports = output_type.get_imports(symbol_table);
        let mut memory_imports = memory.get_imports(symbol_table);
        let mut contract_imports = contract.get_imports(symbol_table);

        // combining all imports and eliminate duplicates
        imports.append(&mut inputs_type_imports);
        imports.append(&mut output_type_imports);
        imports.append(&mut memory_imports);
        imports.append(&mut contract_imports);
        let imports = imports.into_iter().unique().collect::<Vec<_>>();

        // get input's generics: function types in inputs
        let mut generics = inputs
            .iter_mut()
            .flat_map(|(_, typing)| typing.get_generics(&mut identifier_creator))
            .collect::<Vec<_>>();
        let mut output_generics = output_type.get_generics(&mut identifier_creator);
        generics.append(&mut output_generics);

        // get memory/state elements
        let (elements, state_elements_init, state_elements_step) =
            memory.get_state_elements(symbol_table);

        // transform contract
        let contract = contract.lir_from_hir(symbol_table);

        StateMachine {
            name: name.clone(),
            imports,
            input: Input {
                node_name: name.clone(),
                elements: inputs
                    .into_iter()
                    .map(|(identifier, r#type)| InputElement { identifier, r#type })
                    .collect(),
                generics: generics.clone(),
            },
            event,
            state: State {
                node_name: name.clone(),
                elements,
                step: Step {
                    contract,
                    node_name: name.clone(),
                    generics,
                    output_type,
                    body: statements
                        .into_iter()
                        .map(|equation| equation.lir_from_hir(symbol_table))
                        .collect(),
                    state_elements_step,
                    output_expression,
                },
                init: Init {
                    node_name: name,
                    state_elements_init,
                    invariant_initialisation: vec![], // TODO
                },
            },
        }
    }
}
