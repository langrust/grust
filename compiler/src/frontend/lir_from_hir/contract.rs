prelude! {
    hir::contract::Contract,
    lir::contract::Contract as LIRContract,
}

use super::LIRFromHIR;

impl LIRFromHIR for Contract {
    type LIR = LIRContract;

    fn lir_from_hir(self, symbol_table: &SymbolTable) -> Self::LIR {
        let Contract {
            requires,
            ensures,
            invariant,
        } = self;

        LIRContract {
            requires: requires
                .into_iter()
                .map(|term| term.lir_from_hir(symbol_table))
                .collect(),
            ensures: ensures
                .into_iter()
                .map(|term| term.lir_from_hir(symbol_table))
                .collect(),
            invariant: invariant
                .into_iter()
                .map(|term| term.lir_from_hir(symbol_table))
                .collect(),
        }
    }
}

mod term {
    prelude! {
        hir::contract::{Term, term},
    }

    impl super::LIRFromHIR for Term {
        type LIR = lir::contract::Term;

        fn lir_from_hir(self, symbol_table: &SymbolTable) -> Self::LIR {
            match self.kind {
                term::Kind::Constant { constant } => lir::contract::Term::literal(constant),
                term::Kind::Identifier { id } => {
                    let name = symbol_table.get_name(id);
                    match symbol_table.get_scope(id) {
                        Scope::Input => lir::contract::Term::input(name),
                        Scope::Memory => lir::contract::Term::mem(name),
                        Scope::Output => lir::contract::Term::ident("result"), // todo: this will broke for components with multiple outputs
                        Scope::Local => lir::contract::Term::ident(name),
                    }
                }
                term::Kind::Enumeration {
                    enum_id,
                    element_id,
                } => lir::contract::Term::enumeration(
                    symbol_table.get_name(enum_id).clone(),
                    symbol_table.get_name(element_id).clone(),
                    None,
                ),
                term::Kind::Unary { op, term } => {
                    lir::contract::Term::unop(op, term.lir_from_hir(symbol_table))
                }
                term::Kind::Binary { op, left, right } => lir::contract::Term::binop(
                    op,
                    left.lir_from_hir(symbol_table),
                    right.lir_from_hir(symbol_table),
                ),
                term::Kind::ForAll { id, term } => {
                    let name = symbol_table.get_name(id);
                    let ty = symbol_table.get_type(id).clone();
                    let term = term.lir_from_hir(symbol_table);
                    lir::contract::Term::forall(name, ty, term)
                }
                term::Kind::Implication { left, right } => lir::contract::Term::implication(
                    left.lir_from_hir(symbol_table),
                    right.lir_from_hir(symbol_table),
                ),
                term::Kind::PresentEvent { event_id, pattern } => {
                    match symbol_table.get_type(event_id) {
                        Typ::SMEvent { .. } => lir::contract::Term::some(
                            lir::contract::Term::ident(symbol_table.get_name(pattern)),
                        ),
                        Typ::SMTimeout { .. } => {
                            lir::contract::Term::some(lir::contract::Term::ok(
                                lir::contract::Term::ident(symbol_table.get_name(pattern)),
                            ))
                        }
                        _ => unreachable!(),
                    }
                }
                term::Kind::TimeoutEvent { event_id } => match symbol_table.get_type(event_id) {
                    Typ::SMTimeout { .. } => lir::contract::Term::some(lir::contract::Term::err()),
                    _ => unreachable!(),
                },
            }
        }
    }
}
