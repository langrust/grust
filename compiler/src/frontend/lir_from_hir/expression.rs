use itertools::Itertools;
use strum::IntoEnumIterator;

use crate::{
    common::{
        operator::{BinaryOperator, OtherOperator, UnaryOperator},
        r#type::Type,
        scope::Scope,
    },
    hir::expression::{Expression, ExpressionKind},
    lir::{
        block::Block,
        expression::{Expression as LIRExpression, FieldIdentifier},
        item::import::Import,
        pattern::Pattern,
        statement::Statement,
    },
    symbol_table::SymbolTable,
};

use super::LIRFromHIR;

impl<E> LIRFromHIR for ExpressionKind<E>
where
    E: LIRFromHIR<LIR = LIRExpression>,
{
    type LIR = LIRExpression;

    fn lir_from_hir(self, symbol_table: &SymbolTable) -> Self::LIR {
        match self {
            ExpressionKind::Constant { constant, .. } => {
                LIRExpression::Literal { literal: constant }
            }
            ExpressionKind::Identifier { id, .. } => {
                let name = symbol_table.get_name(id).clone();
                if symbol_table.is_function(id) {
                    LIRExpression::Identifier { identifier: name }
                } else {
                    let scope = symbol_table.get_scope(id);
                    match scope {
                        Scope::Input => LIRExpression::InputAccess { identifier: name },
                        Scope::Memory => LIRExpression::MemoryAccess { identifier: name },
                        Scope::Output | Scope::Local => {
                            LIRExpression::Identifier { identifier: name }
                        }
                    }
                }
            }
            ExpressionKind::Unop { op, expression } => {
                let expression = expression.lir_from_hir(symbol_table);
                LIRExpression::Unop {
                    op,
                    expression: Box::new(expression),
                }
            }
            ExpressionKind::Binop {
                op,
                left_expression,
                right_expression,
            } => {
                let left_expression = left_expression.lir_from_hir(symbol_table);
                let right_expression = right_expression.lir_from_hir(symbol_table);
                LIRExpression::Binop {
                    op,
                    left_expression: Box::new(left_expression),
                    right_expression: Box::new(right_expression),
                }
            }
            ExpressionKind::IfThenElse {
                expression,
                true_expression,
                false_expression,
            } => {
                let else_branch = expression.lir_from_hir(symbol_table);
                let then_branch = true_expression.lir_from_hir(symbol_table);
                let condition = false_expression.lir_from_hir(symbol_table);
                LIRExpression::IfThenElse {
                    condition: Box::new(condition),
                    then_branch: Block {
                        statements: vec![Statement::ExpressionLast {
                            expression: then_branch,
                        }],
                    },
                    else_branch: Block {
                        statements: vec![Statement::ExpressionLast {
                            expression: else_branch,
                        }],
                    },
                }
            }
            ExpressionKind::Application {
                function_expression,
                inputs,
                ..
            } => {
                let arguments = inputs
                    .into_iter()
                    .map(|input| input.lir_from_hir(symbol_table))
                    .collect();
                LIRExpression::FunctionCall {
                    function: Box::new(function_expression.lir_from_hir(symbol_table)),
                    arguments,
                }
            }
            ExpressionKind::Abstraction {
                inputs, expression, ..
            } => {
                let inputs = inputs
                    .iter()
                    .map(|id| {
                        (
                            symbol_table.get_name(*id).clone(),
                            symbol_table.get_type(*id).clone(),
                        )
                    })
                    .collect();
                let output = expression.get_type().expect("it should be typed").clone();
                LIRExpression::Lambda {
                    inputs,
                    output,
                    body: Box::new(LIRExpression::Block {
                        block: Block {
                            statements: vec![Statement::ExpressionLast {
                                expression: expression.lir_from_hir(symbol_table),
                            }],
                        },
                    }),
                }
            }
            ExpressionKind::Structure { id, fields, .. } => LIRExpression::Structure {
                name: symbol_table.get_name(id).clone(),
                fields: fields
                    .into_iter()
                    .map(|(id, expression)| {
                        (
                            symbol_table.get_name(id).clone(),
                            expression.lir_from_hir(symbol_table),
                        )
                    })
                    .collect(),
            },
            ExpressionKind::Enumeration { enum_id, elem_id } => LIRExpression::Enumeration {
                name: symbol_table.get_name(enum_id).clone(),
                element: symbol_table.get_name(elem_id).clone(),
            },
            ExpressionKind::Array { elements } => LIRExpression::Array {
                elements: elements
                    .into_iter()
                    .map(|element| element.lir_from_hir(symbol_table))
                    .collect(),
            },
            ExpressionKind::Tuple { elements } => LIRExpression::Tuple {
                elements: elements
                    .into_iter()
                    .map(|element| element.lir_from_hir(symbol_table))
                    .collect(),
            },
            ExpressionKind::Match {
                expression, arms, ..
            } => LIRExpression::Match {
                matched: Box::new(expression.lir_from_hir(symbol_table)),
                arms: arms
                    .into_iter()
                    .map(|(pattern, guard, body, expression)| {
                        (
                            pattern.lir_from_hir(symbol_table),
                            guard.map(|expression| expression.lir_from_hir(symbol_table)),
                            if body.is_empty() {
                                expression.lir_from_hir(symbol_table)
                            } else {
                                let mut statements = body
                                    .into_iter()
                                    .map(|statement| statement.lir_from_hir(symbol_table))
                                    .collect::<Vec<_>>();
                                statements.push(Statement::ExpressionLast {
                                    expression: expression.lir_from_hir(symbol_table),
                                });
                                LIRExpression::Block {
                                    block: Block { statements },
                                }
                            },
                        )
                    })
                    .collect(),
            },
            ExpressionKind::When {
                id,
                option,
                present,
                default,
                ..
            } => LIRExpression::Match {
                matched: Box::new(option.lir_from_hir(symbol_table)),
                arms: vec![
                    (
                        Pattern::Some {
                            pattern: Box::new(Pattern::Identifier {
                                name: symbol_table.get_name(id).clone(),
                            }),
                        },
                        None,
                        present.lir_from_hir(symbol_table),
                    ),
                    (Pattern::None, None, default.lir_from_hir(symbol_table)),
                ],
            },
            ExpressionKind::FieldAccess {
                expression, field, ..
            } => LIRExpression::FieldAccess {
                expression: Box::new(expression.lir_from_hir(symbol_table)),
                field: FieldIdentifier::Named(field),
            },
            ExpressionKind::TupleElementAccess {
                expression,
                element_number,
                ..
            } => LIRExpression::FieldAccess {
                expression: Box::new(expression.lir_from_hir(symbol_table)),
                field: FieldIdentifier::Unamed(element_number),
            },
            ExpressionKind::Map {
                expression,
                function_expression,
                ..
            } => LIRExpression::Map {
                mapped: Box::new(expression.lir_from_hir(symbol_table)),
                function: Box::new(function_expression.lir_from_hir(symbol_table)),
            },
            ExpressionKind::Fold {
                expression,
                initialization_expression,
                function_expression,
                ..
            } => LIRExpression::Fold {
                folded: Box::new(expression.lir_from_hir(symbol_table)),
                initialization: Box::new(initialization_expression.lir_from_hir(symbol_table)),
                function: Box::new(function_expression.lir_from_hir(symbol_table)),
            },
            ExpressionKind::Sort {
                expression,
                function_expression,
                ..
            } => LIRExpression::Sort {
                sorted: Box::new(expression.lir_from_hir(symbol_table)),
                function: Box::new(function_expression.lir_from_hir(symbol_table)),
            },
            ExpressionKind::Zip { arrays, .. } => LIRExpression::Zip {
                arrays: arrays
                    .into_iter()
                    .map(|element| element.lir_from_hir(symbol_table))
                    .collect(),
            },
        }
    }

    fn is_if_then_else(&self, symbol_table: &SymbolTable) -> bool {
        match self {
            ExpressionKind::Identifier { id, .. } => OtherOperator::IfThenElse
                .to_string()
                .eq(symbol_table.get_name(*id)),
            _ => false,
        }
    }

    fn get_imports(&self, symbol_table: &SymbolTable) -> Vec<Import> {
        match self {
            ExpressionKind::Constant { .. } => vec![],
            ExpressionKind::Identifier { id } => {
                if symbol_table.is_function(*id) {
                    if let Some(_) = BinaryOperator::iter()
                        .find(|binary| binary.to_string().eq(symbol_table.get_name(*id)))
                    {
                        vec![]
                    } else if let Some(_) = UnaryOperator::iter()
                        .find(|unary| unary.to_string().eq(symbol_table.get_name(*id)))
                    {
                        vec![]
                    } else if let Some(_) = OtherOperator::iter()
                        .find(|op| op.to_string().eq(symbol_table.get_name(*id)))
                    {
                        vec![]
                    } else {
                        vec![Import::Function(symbol_table.get_name(*id).clone())]
                    }
                } else {
                    vec![]
                }
            }
            ExpressionKind::Unop { expression, .. } => expression.get_imports(symbol_table),
            ExpressionKind::Binop {
                left_expression,
                right_expression,
                ..
            } => {
                let mut imports = left_expression.get_imports(symbol_table);
                let mut right_imports = right_expression.get_imports(symbol_table);
                imports.append(&mut right_imports);
                imports
            }
            ExpressionKind::IfThenElse {
                expression,
                true_expression,
                false_expression,
            } => {
                let mut imports = expression.get_imports(symbol_table);
                let mut true_imports = true_expression.get_imports(symbol_table);
                let mut false_imports = false_expression.get_imports(symbol_table);
                imports.append(&mut true_imports);
                imports.append(&mut false_imports);
                imports
            }
            ExpressionKind::Application {
                function_expression,
                inputs,
            } => {
                let mut imports = function_expression.get_imports(symbol_table);
                let mut inputs_imports = inputs
                    .iter()
                    .flat_map(|expression| expression.get_imports(symbol_table))
                    .unique()
                    .collect::<Vec<_>>();
                imports.append(&mut inputs_imports);
                imports
            }
            ExpressionKind::Abstraction { expression, .. } => expression.get_imports(symbol_table),
            ExpressionKind::Structure { id, fields } => {
                let mut imports = fields
                    .iter()
                    .flat_map(|(_, expression)| expression.get_imports(symbol_table))
                    .unique()
                    .collect::<Vec<_>>();
                imports.push(Import::Structure(symbol_table.get_name(*id).clone()));

                imports
            }
            ExpressionKind::Enumeration { enum_id, .. } => {
                vec![Import::Enumeration(symbol_table.get_name(*enum_id).clone())]
            }
            ExpressionKind::Array { elements } | ExpressionKind::Tuple { elements } => elements
                .iter()
                .flat_map(|expression| expression.get_imports(symbol_table))
                .unique()
                .collect(),
            ExpressionKind::Match { expression, arms } => {
                let mut imports = expression.get_imports(symbol_table);
                let mut arms_imports = arms
                    .iter()
                    .flat_map(|(pattern, option, statements, expression)| {
                        let mut pattern_imports = pattern.get_imports(symbol_table);
                        let mut option_imports = option
                            .as_ref()
                            .map_or(vec![], |expression| expression.get_imports(symbol_table));
                        pattern_imports.append(&mut option_imports);
                        let mut statements_imports = statements
                            .iter()
                            .flat_map(|statement| statement.get_imports(symbol_table))
                            .unique()
                            .collect::<Vec<_>>();
                        pattern_imports.append(&mut statements_imports);
                        let mut expression_imports = expression.get_imports(symbol_table);
                        pattern_imports.append(&mut expression_imports);
                        pattern_imports
                    })
                    .unique()
                    .collect::<Vec<_>>();
                imports.append(&mut arms_imports);
                imports
            }
            ExpressionKind::When {
                option,
                present,
                present_body,
                default,
                default_body,
                ..
            } => {
                let mut imports = option.get_imports(symbol_table);
                let mut present_imports = present.get_imports(symbol_table);
                imports.append(&mut present_imports);
                let mut present_body_imports = present_body
                    .iter()
                    .flat_map(|statement| statement.get_imports(symbol_table))
                    .unique()
                    .collect::<Vec<_>>();
                imports.append(&mut present_body_imports);
                let mut default_imports = default.get_imports(symbol_table);
                imports.append(&mut default_imports);
                let mut default_body_imports = default_body
                    .iter()
                    .flat_map(|statement| statement.get_imports(symbol_table))
                    .unique()
                    .collect::<Vec<_>>();
                imports.append(&mut default_body_imports);
                imports
            }
            ExpressionKind::FieldAccess { expression, .. } => expression.get_imports(symbol_table),
            ExpressionKind::TupleElementAccess { expression, .. } => {
                expression.get_imports(symbol_table)
            }
            ExpressionKind::Map {
                expression,
                function_expression,
            } => {
                let mut imports = expression.get_imports(symbol_table);
                let mut function_imports = function_expression.get_imports(symbol_table);
                imports.append(&mut function_imports);
                imports
            }
            ExpressionKind::Fold {
                expression,
                initialization_expression,
                function_expression,
            } => {
                let mut imports = expression.get_imports(symbol_table);
                let mut initialization_imports =
                    initialization_expression.get_imports(symbol_table);
                imports.append(&mut initialization_imports);
                let mut function_imports = function_expression.get_imports(symbol_table);
                imports.append(&mut function_imports);
                imports
            }
            ExpressionKind::Sort {
                expression,
                function_expression,
            } => {
                let mut imports = expression.get_imports(symbol_table);
                let mut function_imports = function_expression.get_imports(symbol_table);
                imports.append(&mut function_imports);
                imports
            }
            ExpressionKind::Zip { arrays } => arrays
                .iter()
                .flat_map(|expression| expression.get_imports(symbol_table))
                .unique()
                .collect(),
        }
    }
}

impl LIRFromHIR for Expression {
    type LIR = LIRExpression;

    fn lir_from_hir(self, symbol_table: &SymbolTable) -> Self::LIR {
        self.kind.lir_from_hir(symbol_table)
    }
    fn get_type(&self) -> Option<&Type> {
        self.typing.as_ref()
    }

    fn is_if_then_else(&self, symbol_table: &SymbolTable) -> bool {
        self.kind.is_if_then_else(symbol_table)
    }

    fn get_imports(&self, symbol_table: &SymbolTable) -> Vec<Import> {
        self.kind.get_imports(symbol_table)
    }
}
