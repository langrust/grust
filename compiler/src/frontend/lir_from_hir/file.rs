use crate::{
    hir::file::File,
    lir::{item::Item, project::Project},
    symbol_table::SymbolTable,
};

use super::LIRFromHIR;

impl LIRFromHIR for File {
    type LIR = Project;

    fn lir_from_hir(self, symbol_table: &SymbolTable) -> Self::LIR {
        let File {
            typedefs,
            functions,
            nodes,
            interface,
            ..
        } = self;

        let mut typedefs = typedefs
            .into_iter()
            .map(|typedef| typedef.lir_from_hir(symbol_table))
            .collect();
        let mut functions = functions
            .into_iter()
            .map(|function| function.lir_from_hir(symbol_table))
            .map(Item::Function)
            .collect();
        let mut state_machines = nodes
            .into_iter()
            .map(|node| node.lir_from_hir(symbol_table))
            .map(Item::StateMachine)
            .collect();
        let execution_machine = interface.lir_from_hir(symbol_table);

        let mut items = vec![];
        items.append(&mut typedefs);
        items.append(&mut functions);
        items.append(&mut state_machines);
        items.push(Item::ExecutionMachine(execution_machine));

        Project { items }
    }
}
