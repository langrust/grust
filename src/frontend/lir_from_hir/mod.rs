use crate::{common::r#type::Type, lir::item::import::Import, symbol_table::SymbolTable};

/// LIR file construction from HIR project.
pub mod file;

/// LIR function construction from HIR function.
pub mod function;

/// LIR node files construction from HIR node.
pub mod node;

/// LIR expression construction from HIR expression.
pub mod expression;

/// LIR statement construction from HIR statement.
pub mod statement;

/// LIR expression construction from HIR stream expression.
pub mod stream_expression;

/// LIR node file construction from HIR unitary node.
pub mod unitary_node;

/// LIR item construction from HIR typedef.
pub mod typedef;

/// LIR contract construction from HIR contract.
pub mod contract;

/// LIR memory construction from HIR typedef.
pub mod memory;

/// LIR pattern construction from HIR pattern.
pub mod pattern;

/// LIR type construction from HIR type.
pub mod r#type;

/// HIR transformation into LIR.
pub trait LIRFromHIR {
    /// Corresponding LIR construct.
    type LIR;

    /// Transforms HIR into LIR.
    fn lir_from_hir(self, symbol_table: &SymbolTable) -> Self::LIR;

    /// Get type from LIR.
    fn get_type(&self) -> Option<&Type> {
        None
    }

    /// Tell if LIR construct is an IfThenElse operator.
    fn is_if_then_else(&self, _symbol_table: &SymbolTable) -> bool {
        false
    }

    /// Get imports from LIR.
    fn get_imports(&self, _symbol_table: &SymbolTable) -> Vec<Import> {
        vec![]
    }
}
