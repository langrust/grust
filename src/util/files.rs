use codespan_reporting::files;
use std::ops::Range;

/// A file that is backed by an `Arc<String>`.
#[derive(Debug, Clone)]
struct File {
    /// The name of the file.
    name: String,
    /// The source code of the file.
    source: String,
    /// The starting byte indices in the source code.
    line_starts: Vec<usize>,
}

impl File {
    fn line_start(&self, line_index: usize) -> Result<usize, files::Error> {
        use std::cmp::Ordering;

        match line_index.cmp(&self.line_starts.len()) {
            Ordering::Less => Ok(*self
                .line_starts
                .get(line_index)
                .expect("failed despite previous check")),
            Ordering::Equal => Ok(self.source.len()),
            Ordering::Greater => Err(files::Error::LineTooLarge {
                given: line_index,
                max: self.line_starts.len() - 1,
            }),
        }
    }
}

/// An opaque file identifier.
#[derive(Copy, Clone, PartialEq, Eq, Debug)]
pub struct FileId(i32);
impl FileId {
    /// Construct and return the default file identifier.
    pub fn default() -> Self {
        return FileId(-1);
    }
}

/// The files handler.
#[derive(Debug, Clone)]
pub struct Files {
    files: Vec<File>,
}

impl Files {
    /// Create a new files database.
    pub fn new() -> Files {
        Files { files: Vec::new() }
    }

    /// Add a file to the database, returning the handle that can be used to
    /// refer to it again.
    pub fn add(
        &mut self,
        name: impl Into<String>,
        source: impl Into<String>,
    ) -> Option<FileId> {
        let file_id = FileId(i32::try_from(self.files.len()).ok()?);
        let name = name.into();
        let source = source.into();
        let line_starts = files::line_starts(&source).collect();

        self.files.push(File {
            name,
            line_starts,
            source,
        });

        Some(file_id)
    }

    /// Get the file corresponding to the given id.
    fn get(&self, file_id: FileId) -> Result<&File, files::Error> {
        self.files
            .get(file_id.0 as usize)
            .ok_or(files::Error::FileMissing)
    }
}

impl<'files> files::Files<'files> for Files {
    type FileId = FileId;
    type Name = &'files str;
    type Source = &'files str;

    fn name(&self, file_id: FileId) -> Result<&str, files::Error> {
        Ok(self.get(file_id)?.name.as_ref())
    }

    fn source(&self, file_id: FileId) -> Result<&str, files::Error> {
        Ok(&self.get(file_id)?.source)
    }

    fn line_index(
        &self,
        file_id: FileId,
        byte_index: usize
    ) -> Result<usize, files::Error> {
        self.get(file_id)?
            .line_starts
            .binary_search(&byte_index)
            .or_else(|next_line| Ok(next_line - 1))
    }

    fn line_range(
        &self,
        file_id: FileId,
        line_index: usize,
    ) -> Result<Range<usize>, files::Error> {
        let file = self.get(file_id)?;
        let line_start = file.line_start(line_index)?;
        let next_line_start = file.line_start(line_index + 1)?;

        Ok(line_start..next_line_start)
    }
}