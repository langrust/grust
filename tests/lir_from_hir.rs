use codespan_reporting::files::SimpleFiles;

use grustine::lir_from_hir;

#[test]
fn lir_from_hir_transformation_for_counter() {
    let mut files = SimpleFiles::new();

    let counter_id = files.add(
        "counter.gr",
        std::fs::read_to_string("tests/fixture/lir_from_hir/success/counter.gr")
            .expect("unkown file"),
    );

    let project = lir_from_hir(counter_id, &mut files);
    insta::assert_yaml_snapshot!(project)
}

#[test]
fn lir_from_hir_transformation_for_blinking() {
    let mut files = SimpleFiles::new();

    let blinking_id = files.add(
        "blinking.gr",
        std::fs::read_to_string("tests/fixture/lir_from_hir/success/blinking.gr")
            .expect("unkown file"),
    );

    let project = lir_from_hir(blinking_id, &mut files);
    insta::assert_yaml_snapshot!(project)
}

#[test]
fn lir_from_hir_transformation_for_button_management() {
    let mut files = SimpleFiles::new();

    let button_management_id = files.add(
        "button_management.gr",
        std::fs::read_to_string("tests/fixture/lir_from_hir/success/button_management.gr")
            .expect("unkown file"),
    );

    let project = lir_from_hir(button_management_id, &mut files);
    insta::assert_yaml_snapshot!(project)
}

#[test]
fn lir_from_hir_transformation_for_button_management_condition_match() {
    let mut files = SimpleFiles::new();

    let button_management_condition_match_id = files.add(
        "button_management_condition_match.gr",
        std::fs::read_to_string(
            "tests/fixture/lir_from_hir/success/button_management_condition_match.gr",
        )
        .expect("unkown file"),
    );

    let project = lir_from_hir(button_management_condition_match_id, &mut files);
    insta::assert_yaml_snapshot!(project)
}

#[test]
fn lir_from_hir_transformation_for_button_management_using_function() {
    let mut files = SimpleFiles::new();

    let button_management_using_function_id = files.add(
        "button_management_using_function.gr",
        std::fs::read_to_string(
            "tests/fixture/lir_from_hir/success/button_management_using_function.gr",
        )
        .expect("unkown file"),
    );

    let project = lir_from_hir(button_management_using_function_id, &mut files);
    insta::assert_yaml_snapshot!(project)
}

#[test]
fn lir_from_hir_transformation_for_pid() {
    let mut files = SimpleFiles::new();

    let pid_id = files.add(
        "pid.gr",
        std::fs::read_to_string("tests/fixture/lir_from_hir/success/pid.gr").expect("unkown file"),
    );

    let project = lir_from_hir(pid_id, &mut files);
    insta::assert_yaml_snapshot!(project)
}
