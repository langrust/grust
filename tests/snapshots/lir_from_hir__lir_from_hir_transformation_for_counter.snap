---
source: tests/lir_from_hir.rs
expression: project
---
items:
  - NodeFile:
      name: counter_o
      imports: []
      input:
        node_name: counter_o
        elements:
          - identifier: res
            type: Boolean
          - identifier: tick
            type: Boolean
      state:
        node_name: counter_o
        elements:
          - Buffer:
              identifier: mem_o
              type: Integer
        step:
          node_name: counter_o
          output_type: Integer
          body:
            - Let:
                identifier: inc
                expression:
                  IfThenElse:
                    condition:
                      InputAccess:
                        identifier: tick
                    then_branch:
                      statements:
                        - ExpressionLast:
                            expression:
                              Literal:
                                literal:
                                  Integer: 1
                    else_branch:
                      statements:
                        - ExpressionLast:
                            expression:
                              Literal:
                                literal:
                                  Integer: 0
            - Let:
                identifier: o
                expression:
                  IfThenElse:
                    condition:
                      InputAccess:
                        identifier: res
                    then_branch:
                      statements:
                        - ExpressionLast:
                            expression:
                              Literal:
                                literal:
                                  Integer: 0
                    else_branch:
                      statements:
                        - ExpressionLast:
                            expression:
                              FunctionCall:
                                function:
                                  Identifier:
                                    identifier: " + "
                                arguments:
                                  - FunctionCall:
                                      function:
                                        Identifier:
                                          identifier: (_)
                                      arguments:
                                        - MemoryAccess:
                                            identifier: mem_o
                                  - Identifier:
                                      identifier: inc
          state_elements_step:
            - identifier: mem_o
              expression:
                Identifier:
                  identifier: o
          output_expression:
            Identifier:
              identifier: o
        init:
          node_name: counter_o
          state_elements_init:
            - BufferInit:
                identifier: mem_o
                initial_value:
                  Integer: 0
  - NodeFile:
      name: main_y
      imports:
        - NodeFile: counter_o
      input:
        node_name: main_y
        elements: []
      state:
        node_name: main_y
        elements:
          - Buffer:
              identifier: mem_half
              type: Boolean
          - Buffer:
              identifier: mem_x
              type: Boolean
          - CalledNode:
              identifier: counter_o_y
              node_name: counter_o
        step:
          node_name: main_y
          output_type: Integer
          body:
            - Let:
                identifier: half
                expression:
                  MemoryAccess:
                    identifier: mem_half
            - Let:
                identifier: x
                expression:
                  MemoryAccess:
                    identifier: mem_x
            - LetTuple:
                identifiers:
                  - counter_o_y
                  - y
                expression:
                  NodeCall:
                    node_identifier: counter_o_y
                    input_name: CounterOInput
                    input_fields:
                      - - res
                        - Identifier:
                            identifier: x
                      - - tick
                        - Identifier:
                            identifier: half
          state_elements_step:
            - identifier: mem_half
              expression:
                FunctionCall:
                  function:
                    Identifier:
                      identifier: "!"
                  arguments:
                    - Identifier:
                        identifier: half
            - identifier: mem_x
              expression:
                FunctionCall:
                  function:
                    Identifier:
                      identifier: (_)
                  arguments:
                    - FunctionCall:
                        function:
                          Identifier:
                            identifier: " > "
                        arguments:
                          - Identifier:
                              identifier: y
                          - Literal:
                              literal:
                                Integer: 35
            - identifier: counter_o_y
              expression:
                Identifier:
                  identifier: counter_o_y
          output_expression:
            Identifier:
              identifier: y
        init:
          node_name: main_y
          state_elements_init:
            - BufferInit:
                identifier: mem_half
                initial_value:
                  Boolean: true
            - BufferInit:
                identifier: mem_x
                initial_value:
                  Boolean: false
            - CalledNodeInit:
                identifier: counter_o_y
                node_name: counter_o

